sw/device/tock/chips/earlgrey/src/device_config.rs

```diff
@@ -0,0 +1,22 @@
+//! Device specific configuration.
+
+pub struct DeviceConfig<'a> {
+    pub name: &'a str,
+    pub chip_freq: u32,
+    pub uart_baudrate: u32,
+}
+
+// This is also a default case, when no device is specified.
+#[cfg(any(feature = "device_fpga", not(feature = "device_specified")))]
+pub const DEVICE_CONFIG: DeviceConfig = DeviceConfig {
+    name: &"device_fpga",
+    chip_freq: 50_000_000,
+    uart_baudrate: 230400,
+};
+
+#[cfg(feature = "device_verilator")]
+pub const DEVICE_CONFIG: DeviceConfig = DeviceConfig {
+    name: &"device_verilator",
+    chip_freq: 500_000,
+    uart_baudrate: 9600,
+};
```

Comment:
Hey @alphan , thank you for your review.

> But, I think it would be better if you could add a definition that fails to compile for the the case when no device is specified (instead of silently using device_fpga).

Can do, the reason why I have done it this way - is to do with upstream CI, which builds every crate separately. This means that we can't put `default` feature in the crate (default feature is always built, and will conflict if we pass another option), so essentially I have implemented the default case in code instead. For downstream, we can make it fail - but then effectively the default case moves to the point of `build_tock.sh` or `meason.build` level. I think these improvements should be done at the later stage/follow-up (the second stage #2139 ).

> Also, can you match these to the names used in sw/device/lib/arch/meson.build, i.e. sim_verilator and fpga_nexysvideo.

Sure, will do.

> I find it a bit counterintuitive to name the chip-specific directory earlgrey and the shared-code directory opentitan_common.

I can see that, but in my opinion it is not an unreasonable split, I see OpenTitan as a project, and earlgrey a specific chip implementation of this project. It is true that we don't need this split at the moment, as we only have one chip.

> If there are no other crates that depend on opentitan_common, maybe we should hold off on this split until we need it.

This is essentially what the upstream `ibex` and `lowrisc` crates are, I am happy to rename `opentitan_common` to something better. Initially I had it names `peripherals`, but then that doesn't work if we eventually decide to upstream it.

The only alternative I see, is to stick `opentitan_common` under `earlgrey`, either as a create or just put peripheral `.rs` files there. As they are more or less copies of the `ibex` and `lowrisc` crates in upstream, it would be easier to the merge (if necessary) in the second stage (#2139 ).

> Is there a reason why the code for plic and timer is not moved to opentitan_common
This is how these peripherals were implemented in the upstream, and further peripheral refactoring is subject to the second stage of the Tock downstream development (#2139 )

The idea is to take smaller steps at the time, there are plenty of improvements and tidying up that can be done in follow-ups.